api/src/pcapi/core/users/email/update.py

```diff
@@ -104,39 +203,57 @@ def full_email_update_by_admin(user: models.User, email: str) -> None:
     repository.save(user, admin_update_event)
 
 
-def get_no_active_token_key(user: models.User) -> str:
-    return f"update_email_active_tokens_{user.id}"
+def get_token_key(user: models.User, token_type: TokenType) -> str:
+    match token_type:
+        case TokenType.CONFIRMATION:
+            return f"update_email_confirmation_token_{user.id}"
+        case TokenType.VALIDATION:
+            return f"update_email_validation_token_{user.id}"
+        case _:
+            raise ValueError(f"Invalid token type: {token_type}")
 
 
-def check_no_active_token_exists(user: models.User, expiration_date: datetime) -> None:
-    """
-    Use a dummy counter to find out whether the user already has an
-    active token.
+def token_exists(user: models.User, token_type: TokenType) -> bool:
+    """Check if there is an active confirmation token for the user"""
+    key = get_token_key(user, token_type)
+    return app.redis_client.exists(key)  # type: ignore [attr-defined]
 
-    * If the incr command returns 1, there were none. Hence, set a TTL
-      (expiration_date, the lifetime of the validation token).
-    * If not, raise an error because there is already one.
-    """
-    key = get_no_active_token_key(user)
-    count = app.redis_client.incr(key)  # type: ignore [attr-defined]
 
-    if count > 1:
-        raise exceptions.EmailUpdateTokenExists()
+def generate_email_change_token(
+    user: models.User, new_email: str, expiration_date: datetime, token_type: TokenType
+) -> str:
+    token = encode_jwt_payload({"current_email": user.email, "new_email": new_email}, expiration_date)
+    app.redis_client.set(get_token_key(user, token_type), token)  # type: ignore [attr-defined]
+    app.redis_client.expireat(get_token_key(user, token_type), expiration_date)  # type: ignore [attr-defined]
+    return token
 
-    app.redis_client.expireat(key, expiration_date)  # type: ignore [attr-defined]
 
+def check_and_expire_token(user: models.User, token: str, token_type: TokenType) -> None:
+    token_key = get_token_key(user, token_type)
+    if app.redis_client.get(token_key) == token:  # type: ignore [attr-defined]
+        app.redis_client.delete(token_key)  # type: ignore [attr-defined]
+    else:
+        raise exceptions.InvalidToken()
 
-def get_active_token_expiration(user) -> datetime | None:  # type: ignore [no-untyped-def]
-    key = get_no_active_token_key(user)
-    ttl = app.redis_client.ttl(key)  # type: ignore [attr-defined]
+
+def get_active_token_expiration(user: models.User) -> datetime | None:
+    key = get_token_key(user, TokenType.CONFIRMATION)
+    if app.redis_client.exists(key):  # type: ignore [attr-defined]
+        ttl = app.redis_client.ttl(key)  # type: ignore [attr-defined]
+    else:
+        key = get_token_key(user, TokenType.VALIDATION)
+        if app.redis_client.exists(key):  # type: ignore [attr-defined]
+            ttl = app.redis_client.ttl(key)  # type: ignore [attr-defined]
+        else:
+            ttl = -1
```

Comment:
Aie j'avais pas vu que c'est toujours là ça !
<!-- Issues represent user-facing problems. If possible, it's great to follow this kind of comment with a suggestion.  -->
**🔨 issue:** ‏80% sur que ça marche pas quand on en est à la validation